### ISSUE TITLE
Crashing with Multicore in cax.process.py on Midway

### ISSUE THREAD

**Issue by pdeperio** [XENON1T/pax#351]
Jobs submitted via cax.process.py on Midway using 4 cores crashes:

https://gist.github.com/pdeperio/44281f9deb8c680d750c22fc48ab4909
Dataset: 160501_1331
Run: 338

https://gist.github.com/pdeperio/67b715e74ca2c95e89599f374771d222
Dataset: 160501_1314
Run: 337

etc...

**Comment by pdeperio** [comment]
Another strange ending which may be a bit more dangerous since it looks like it completed happily:

https://gist.github.com/pdeperio/5ff2dcf22b4f94f613d1fc993c95372c

**Comment by pdeperio** [comment]
Still crashing even with 1 cpu:

https://gist.github.com/pdeperio/041a4581a834986d19ee9e8b96902a42

```AttributeError: 'float' object has no attribute 'typecode'```

**Comment by pdeperio** [comment]
This job looks like it completed (ROOT file seems OK), but there's an error within ```TypeError: 'int' object is not callable```:

https://gist.github.com/pdeperio/775fab31c5f46b547a2ba17fb207cdcc

and the RunsDB entry seems to be stuck at "transferring", instead of "transferred" or "error"

**Comment by JelleAalbers** [comment]
These seem to be very different errors. Can you reproduce them? i.e. if you just run pax on one of those root files, does the error recur?

If not, one possibility is the class sometimes did not finish compiling when the other workers try to load it: I was to lazy to implement a lock-based solution for this race condition (https://github.com/XENON1T/pax/blob/master/pax/plugins/io/ROOTClass.py#L376). Perhaps you can try increasing the value here: https://github.com/XENON1T/pax/blob/master/pax/plugins/io/ROOTClass.py#L382 to some ridiculously high number, and see if it works better. Or perhaps I should just implement a proper locking solution :-)

**Comment by pdeperio** [comment]
Can this issue be exacerbated by the several jobs starting simultaneously in parallel?

**Comment by ershockley** [comment]
I processed these same datasets to see if I saw the same error: 160501_1314 crashed similarly to what Patrick saw, but 160501_1331 finished successfully after about 3 hours. I submitted both of these jobs at the same time. 

crash: https://gist.github.com/ershockley/21744c57f11fffb8b002b8978ff2328f
success: https://gist.github.com/ershockley/51e166216493795b5a512979ad8c320e

I tried 160501_1314 again, and it's been running for about 45 minutes now without crashing. I'll check if it completed it in the morning.

**Comment by JelleAalbers** [comment]
Thanks for these reports!

I just released 4.9.2, this should solve these issues if / insofar as they are due to a race condition in the pax ROOT event class compilation. The process compiling the pax class now creates a lock file; the other processes wait for it to disappear before they try to load the class. 

Perhaps it is a good idea to clear the rootpy cache (`~/.cache/rootpy`) once in a while to remove any failed compilation residues. If you do all your compilation in the same folder, try clearing the pax_event_class* files after a crash to ensure no mess remains.

I'm not sure whether you do this at your setup, but it is probably not safe to start several pax instances in the same folder. If you have bad luck, both may simultaneously find no lock file present and start compilation, with unpredictable results.

**Comment by pdeperio** [comment]
Thanks Jelle, jobs seem to be completing successfully now, except for one that seems to be stuck at:

```Status: input_done. Processing queue: 0 events. Output queue: 0 events. RAM usage: 352.8 (master) 1047.4 (workers) 379.8 (output)```

```
Dataset: 160503_1651
Run: 415
Log: /project/lgrandi/xenon1t/processing/logs/160503_1651_18809152.log 
```
